### PR TITLE
OCC startup (istep 21.1)

### PR DESCRIPTION
**Pull request by SergiiDmytruk**
Includes #89.

### CODE REVIEW

File: src/include/cpu/power/occ.h
Comment (krystian-hebel) [inline src/include/cpu/power/occ.h]: These are wrong, should be 0x48000002 and 0x03FFFFFC respectively, this error was copied from Hostboot. Nonetheless, it is not actually used so can be removed.

File: src/include/cpu/power/occ.h
Comment (krystian-hebel) [inline src/include/cpu/power/occ.h]: There are 12 EXs, which are accessed through quad chiplets. Names are misleading, as well as use of this table in clear_occ_special_wakeups() - see comment there.

File: src/include/cpu/power/occ.h
Comment (krystian-hebel) [inline src/include/cpu/power/occ.h]: Unused, please remove.

File: src/include/cpu/power/occ.h
Comment (krystian-hebel) [inline src/include/cpu/power/occ.h]: AFAICT only OCB0 is ever used, drop these tables and use appropriate register directly.

File: src/soc/ibm/power9/Makefile.inc
Comment (krystian-hebel) [inline src/soc/ibm/power9/Makefile.inc]: Why is it needed for romstage?

File: src/soc/ibm/power9/occ.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/occ.c]: As mentioned in the header, there are 12 EXs, which are accessed through even core chiplets. This loop should be something like this:

```
for (size_t chiplet_index = 0; chiplet_index < MAX_CORES /* 24 */; chiplet_index += 2) {
    write_scom_for_chiplet(EP00_CHIPLET_ID + chiplet_index, EX_PPM_SPWKUP_OCC, ....
}
```
or
```
for (size_t ex_index = 0; ex_index < MAX_EXS /* 12 */; ++ex_index) {
    write_scom_for_chiplet(EP00_CHIPLET_ID + ex_index * 2, EX_PPM_SPWKUP_OCC, ....
}
```

File: src/include/cpu/power/occ.h
Comment (krystian-hebel) [reply src/include/cpu/power/occ.h]: Also, if any register address is not needed outside of occ.c, move its definition there.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: According to dumps, this is 8M for BAR2 (0x700000). This is also set [here](https://github.com/3mdeb/coreboot/blob/a11f8a632a089c3de818f467f6d6abc658222588/src/soc/ibm/power9/homer.c#L2128) - one of these should be enough.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Hostboot reserves memory for 8 HOMER images (4M each) and one common area (8M) after the last one. For Talos we will need two HOMERs when second CPU is installed, so it shouldn't be hardcoded. Second CPU will also use different address of HOMER, but the same common area, so for now it will be enough to just pass pointer to common_area to load_pm_complex() and hardcode it at the highest possible level.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: This explains last unaccounted differences between our HOMER and one prepared by Hostboot somewhere after 15.2

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Is it supposed to work on cores, and not EXs as clear_occ_special_wakeups() did? I don't see this in dumps at all.

The final read is some kind of workaround? If yes, please add a comment.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Use `PU_OCB_PIB_OCBCSR0_OR` directly.

File: src/soc/ibm/power9/occ.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/occ.c]: If this is always for OCB0, use register names directly.

File: src/soc/ibm/power9/occ.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/occ.c]: Does it work without this delay? Maybe it was used for simulation only.

File: src/soc/ibm/power9/occ.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/occ.c]: Unused, please remove.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Maybe use 0x7C0003A6 as MTSPR_OP and drop this? This is extended opcode, without it instruction isn't `mtspr` anyway.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Same as above.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: BR_OP is already shifted, so:
```suggestion
	inst = BR_OP;
```
No idea why it works in this form...

**Review by krystian-hebel — CHANGES_REQUESTED** [review]

File: src/soc/ibm/power9/Makefile.inc
Comment (SergiiDmytruk) [reply src/soc/ibm/power9/Makefile.inc]: This is from #89, I removed this line in a later commit.

File: src/soc/ibm/power9/occ.c
Comment (SergiiDmytruk) [reply src/soc/ibm/power9/occ.c]: Haven't tried without the delay, will test. I'm surprised it works at all... bits are checked with `& bit_pos` instead of `& PPC_BIT(bit_pos)`. Haven't looked at it closely before.

File: src/soc/ibm/power9/homer.c
Comment (SergiiDmytruk) [reply src/soc/ibm/power9/homer.c]: This is from an old commit.

File: src/include/cpu/power/occ.h
Comment (SergiiDmytruk) [reply src/include/cpu/power/occ.h]: Moved more stuff to `occ.c`, which simplified things. I was just expecting #89 to be merged first and avoided modifying `occ.c` too much.

File: src/soc/ibm/power9/homer.c
Comment (SergiiDmytruk) [reply src/soc/ibm/power9/homer.c]: I probably did iteration over cores because they are used for computing address in `src/import/chips/p9/procedures/hwp/pm/p9_cpu_special_wakeup_ex.C` (code related to wakeups is rather convoluted) and logs have this as if they iterate over cores as well:
```
553.73054|SCAN|TRACE : PUTSCOM     :  pu:k0:n0:s0:p00 : 00000000210F010B 0000000000000000
553.73056|SCAN|TRACE : GETSCOM     :  pu:k0:n0:s0:p00 : 00000000210F010B 0000000000000000

553.73084|SCAN|TRACE : PUTSCOM     :  pu:k0:n0:s0:p00 : 00000000230F010B 0000000000000000
553.73086|SCAN|TRACE : GETSCOM     :  pu:k0:n0:s0:p00 : 00000000230F010B 0000000000000000

553.73114|SCAN|TRACE : PUTSCOM     :  pu:k0:n0:s0:p00 : 00000000240F010B 0000000000000000
553.73116|SCAN|TRACE : GETSCOM     :  pu:k0:n0:s0:p00 : 00000000240F010B 0000000000000000

553.73145|SCAN|TRACE : PUTSCOM     :  pu:k0:n0:s0:p00 : 00000000270F010B 0000000000000000
553.73147|SCAN|TRACE : GETSCOM     :  pu:k0:n0:s0:p00 : 00000000270F010B 0000000000000000
```
Hostboot seems to send wakeup to functional core of an EX instead of the first one.

Final read adds a delay, copied comment from Hostboot.

File: src/soc/ibm/power9/occ.c
Comment (SergiiDmytruk) [reply src/soc/ibm/power9/occ.c]: Removed the delay. Seems to make no difference.

File: src/soc/ibm/power9/Makefile.inc
Comment (krystian-hebel) [reply src/soc/ibm/power9/Makefile.inc]: Huh, I have no idea how but I managed to review older version, sorry about that.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Here and for all other SCOMs: I'd rather see PPC_BIT(MASK) macros used for this, as long as we know the layout of the register. That way it is easier to understand and check if code does what comment says without manually counting bit positions which is error prone, especially when value doesn't have all 16 hex-digits. Consecutive fields can be covered with one macro, e.g. PPC_BITMASK(0,14) in this case.

If that is not possible, please use capital letters for hex values.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]:
```suggestion
			err_mask = 0xFFF00000; // from Hostboot's log
```

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]:
```suggestion
	wait_ms(PGPE_POLLTIME_MS * TIMEOUT_COUNT,
```

I wonder if we should add a version of wait_ms() to coreboot that checks condition periodically instead of using busy loop. For memory training busy reading significantly slowed down the training. The issue is that we can't use delay() in ramstage, at least not in the same form. Still, this is not part of this task, I'm just thinking out loud :slightly_smiling_face:

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Is this correct? PBA_OCI_MARKER_BASE is 0x40070000, it won't fit in just 3 bits. Maybe just set it to 0x7 and skip shifting.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: This is almost identical to the last lines of pba_reset(), except write gather is disabled for slave 0 there. As SGPE isn't started before pba_slave_setup_runtime_phase() anyway, could you remove lines 790-804? You did much better work describing those registers than me.

Comment (krystian-hebel) [reply src/soc/ibm/power9/homer.c]: Also, as these are not documented, please copy the description from comments below to [undocumented_registers.md](https://github.com/3mdeb/openpower-coreboot-docs/blob/main/devnotes/undocumented_registers.md)

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]:
```suggestion
			*conf_vector |= PPC_BIT(pos);
```
Here and below

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Please add comment about which bits are set. Seems that some of them should be set with PPC_PLACE/PPC_SHIFT instead.

File: src/soc/ibm/power9/homer.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/homer.c]: Is it even possible? parse_occ_response() checks for OCC_RC_SUCCESS and returns false, so previous `if` will be taken in that case.

File: src/soc/ibm/power9/occ.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/occ.c]: This line is useless, `*_for_chiplet()` functions already add proper offset. Remove it and use `PPM_SPWKUP_FSP` instead of `spwkup_addr`

File: src/soc/ibm/power9/occ.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/occ.c]: Please make those inline and move to ops.h. I can think of at least one use for ppc_mtspr elsewhere.

**Review by krystian-hebel — CHANGES_REQUESTED** [review]

File: src/include/cpu/power/occ.h
Comment (krystian-hebel) [reply src/include/cpu/power/occ.h]: We will have to divide the code even more for upstreaming and probably move the rest of OCC functions to `occ.c`, but for now this should be good enough.

File: src/soc/ibm/power9/homer.c
Comment (SergiiDmytruk) [reply src/soc/ibm/power9/homer.c]: That's how Hostboot does it. I guess `0x40070000` is a final address and `oci_marker_space` is an offset in 64 KiB units from a base address. But yes, it looks too complicated.

File: src/soc/ibm/power9/homer.c
Comment (SergiiDmytruk) [reply src/soc/ibm/power9/homer.c]: Ended up calling this function once on line 790. Calling it only in `pm_pba_init()` caused hang somewhere in `start_pm_complex()`.

File: src/soc/ibm/power9/fir.c
Comment (krystian-hebel) [inline src/soc/ibm/power9/fir.c]: Almost each component has it's own set of FIR, having this in separate file is misleading. Please move it to occ.c. Same goes for fir.h.

**Review by krystian-hebel — CHANGES_REQUESTED** [review]
One small change, otherwise LGTM

**Review by krystian-hebel — APPROVED** [review]

**Comment by krystian-hebel** [comment]
Let's wait for CI tests before merging.

**Pull request merged**
